redeem/boards/replicape.py

```diff
@@ -0,0 +1,163 @@
+import logging
+import glob
+import struct
+
+from redeem.PWM import PWM
+from redeem.Enable import Enable
+from redeem.EndStop import EndStop
+from redeem.Stepper import Stepper_00A3, Stepper_00A4, Stepper_00B1, Stepper_00B2, Stepper_00B3
+from redeem.Key_pin import Key_pin, Key_pin_listener
+from redeem.Mosfet import Mosfet
+from redeem.Fan import Fan
+
+
+def probe_replicape(printer):
+  results = find_replicape()
+  if results == None:
+    logging.warn("No replicape found")
+    return
+
+  eeprom_path, eeprom_data, revision = results
+  replicape_key = get_replicape_key(eeprom_path, eeprom_data)
+  build_replicape_printer(printer, revision)
+
+
+def find_replicape():
+  found = False
+  revision = None
+  eeprom_data = None
+  eeprom_path = None
+
+  nvmem_paths = glob.glob("/sys/bus/nvmem/devices/[1-2]-005[4-7]0/nvmem")
+  for i, nvmem_path in enumerate(nvmem_paths):
+    try:
+      with open(nvmem_path, "rb") as f:
+        data = f.read(120)
+        name = data[58:74].strip()
+        if name == "BB-BONE-REPLICAP":
+          revision = data[38:42]
+          eeprom_data = data
+          eeprom_path = nvmem_path
+          found = True
+          break
+    except IOError as e:
+      pass
+
+  if not found:
+    return None
+
+  logging.debug("Found replicape revision %s at %s", revision, eeprom_path)
+  return eeprom_path, eeprom_data, revision
+
+
+def get_replicape_key(eeprom_path, eeprom_data):
+  """ Get the generated key from the config or create one """
+  replicape_key = "".join(struct.unpack('20c', eeprom_data[100:120]))
+  logging.debug("Found Replicape key: '" + replicape_key + "'")
+  if replicape_key == '\x00' * 20:
+    logging.debug("Replicape key invalid")
+    import random
+    import string
+    replicape_key = ''.join(
+        random.SystemRandom().choice(string.ascii_uppercase + string.digits) for _ in range(20))
+    eeprom_data = eeprom_data[:100] + replicape_key
+    logging.debug("New Replicape key: '" + replicape_key + "'")
+    try:
+      with open(eeprom_path, "wb") as f:
+        f.write(eeprom_data[:120])
+    except IOError as e:
+      logging.warning("Unable to write new key to EEPROM")
+  return replicape_key
+
+
+def build_replicape_printer(printer, revision):
```

Comment:
I wonder if this would benifit from creating a new "Replicape_HW" base class, and then inherit from it for the B3, B2, B1, A4, A3, ect. Then we could put general logic in the base class and then only override what is needed in the child classes.
I would probably do multiple levels:
A4 inherits from A inherits from Redeem_HW
That would then allow us to easily add new hw  boards, and even create the virtual hw board without much issue at all.

Reply:
Note: this suggestion is similar to what was done with the stepper class, though there is some inheritance "problems" where the inheritance is not being fully utilized in that code.

Reply:
I did consider this, but I ultimately decided that the differences were small enough and the overall code was compact enough that it was better to keep it all together to make the differences between revisions more obvious. I suspect this is mostly a matter of taste.

This also won't be an extension point because there are unlikely to be more Replicape revisions.